### PR TITLE
TL/UCP: two tree pattern and bcast

### PR DESCRIPTION
**Pull request by shimmybalsam**
## What
- Implementation of the double binary tree pattern, including both options: mirror (for even size team) and shift (for odd size team).
- Implementation of bcast using double binary tree.

## Why ?
First step for future double binary tree allreduce (reduce+bcast)

## How ?
Building the first tree, T1, is done in an "in order" binary tree fashion. 
To build T2 we use either shift for odd size team, which is a "shift left by 1" of T1, or mirroring for even size team which is an exact mirrored image of T1 graph, while re numbering the ranks to keep "in order".
At the end of both shift and mirror options, the leaf nodes from T1 will be intermediate (or root) nodes in T2, allowing for better bidirectional BW usage.

Based on alg paper: http://algo2.iti.kit.edu/documents/2tree.pdf

For bcast:
We split the data of the root in 2 and send first half throughout T1 and second half throughout T2.

### CODE REVIEW

**Review by samnordmann — COMMENTED** [review]
Looks really good, thanks @shimmybalsam !

File: src/components/tl/ucp/bcast/bcast_two_tree.c [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    ucc_datatype_t              dt        = TASK_ARGS(task).src.info.datatype;
+    size_t                      count     = TASK_ARGS(task).src.info.count;
+    size_t                      data_size = count * ucc_dt_size(dt) / 2;
+    ucc_rank_t                  coll_root = (ucc_rank_t)TASK_ARGS(task).root;
+    ucc_rank_t                  t1_root   = task->bcast_two_tree.t1.root;
+    ucc_rank_t                  t2_root   = task->bcast_two_tree.t2.root;
+    ucp_tag_recv_nbx_callback_t cb[2]     = {recv_completion_1,
+                                             recv_completion_2};
+
+    task->bcast_two_tree.t1.recv = 0;
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/bcast/bcast_two_tree.c:163]: needed for persistent

File: test/gtest/coll/test_bcast.cc [hunk 2]
```diff
@@ -280,3 +280,77 @@ UCC_TEST_F(test_bcast_alg, 2step) {
         }
     }
 }
+
+UCC_TEST_F(test_bcast_alg, two_tree_odd_shift) {
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply test/gtest/coll/test_bcast.cc:284]: We build T2 differently (shift/mirror) if we for even and odd size team so in my opinion they should be separate independent tests.

[the remark this answers is not in the document]
Comment (samnordmann) [reply test/gtest/coll/test_bcast.cc:284]: I am not sure to understand. They will still be separate tests, just te implementation will be factorized and we will avoid duplication

File: src/components/tl/ucp/bcast/bcast_two_tree.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (samnordmann) [reply src/components/tl/ucp/bcast/bcast_two_tree.c:163]: so it is needed to have them both in `ucc_two_tree_build_trees` and here ?

File: src/components/tl/ucp/bcast/bcast.c [hunk 3]
```diff
@@ -19,6 +19,11 @@ ucc_base_coll_alg_info_t
              .name = "sag_knomial",
              .desc = "recursive knomial scatter followed by knomial "
                      "allgather (optimized for BW)"},
+        [UCC_TL_UCP_BCAST_ALG_TWO_TREE] =
+            {.id   = UCC_TL_UCP_BCAST_ALG_TWO_TREE,
+             .name = "two_tree",
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/bcast/bcast.c]: changed to dbt

File: src/components/tl/ucp/tl_ucp_coll.h [hunk 4]
```diff
@@ -194,6 +200,11 @@ typedef struct ucc_tl_ucp_task {
             ucc_ee_executor_task_t *etask;
             ucc_ee_executor_t      *executor;
         } reduce_kn;
+        struct {
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/tl_ucp_coll.h]: yes , removed

File: src/Makefile.am [hunk 5]
```diff
@@ -91,6 +91,7 @@ noinst_HEADERS =                      \
 	coll_patterns/recursive_knomial.h \
 	coll_patterns/sra_knomial.h       \
 	coll_patterns/bruck_alltoall.h    \
+	coll_patterns/two_tree.h          \
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/Makefile.am]: changed to double_binary_tree

File: src/components/tl/ucp/bcast/bcast_two_tree.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply src/components/tl/ucp/bcast/bcast_two_tree.c]: it is needed in build_trees function to not have coverity initialization error

File: test/gtest/coll/test_bcast.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply test/gtest/coll/test_bcast.cc]: merged tests

**Comment by shimmybalsam** [comment]
Compariosn of bcast SAG vs DBT:
If opt radix in SAG gives small amount of extra ranks then it performs best, otherwise, with "bad" radix with a lot of extra ranks, DBT is better.

image

**Review by samnordmann — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]
merging as is since @shimmybalsam currently can't work on this PR

**Comment by pangjuan8848** [comment]
Hello, I am interested in this paper, but the link is broken, where else can I read this paper？
 alg paper: http://algo2.iti.kit.edu/documents/2tree.pdf

**Comment by Sergei-Lebedev** [comment]
> Hello, I am interested in this paper, but the link is broken, where else can I read this paper？ alg paper: http://algo2.iti.kit.edu/documents/2tree.pdf

Try to search for "Two-tree algorithms for full bandwidth broadcast, reduction and scan" in scholar.google.com

**Comment by pangjuan8848** [comment]
> > Hello, I am interested in this paper, but the link is broken, where else can I read this paper？ alg paper: http://algo2.iti.kit.edu/documents/2tree.pdf
> 
> Try to search for "Two-tree algorithms for full bandwidth broadcast, reduction and scan" in scholar.google.com

oh, thank you, i got it

**Pull request merged**